From e5d4a116af23561e21006f5096f37b06f3991bba Mon Sep 17 00:00:00 2001
From: Michael Gehring <mg@ebfe.org>
Date: Sun, 25 Aug 2013 19:24:27 +0200
Subject: [PATCH] continuous scrolling

---
 platform/x11/pdfapp.c | 36 +++++++++++++++++++++++++++++++-----
 1 file changed, 31 insertions(+), 5 deletions(-)

diff --git a/platform/x11/pdfapp.c b/platform/x11/pdfapp.c
index 34e0a8c..86fd6d1 100644
--- a/platform/x11/pdfapp.c
+++ b/platform/x11/pdfapp.c
@@ -1112,13 +1112,29 @@ void pdfapp_onkey(pdfapp_t *app, int c)
 		break;
 
 	case 'j':
-		app->pany -= fz_pixmap_height(app->ctx, app->image) / 10;
-		pdfapp_showpage(app, 0, 0, 1, 0);
+		{
+			int image_h = fz_pixmap_height(app->ctx, app->image);
+			if (app->pany + image_h <= app->winh || app->winh >= image_h) {
+				panto = PAN_TO_TOP;
+				app->pageno++;
+				break;
+			}
+			app->pany -= image_h / 10;
+			pdfapp_showpage(app, 0, 0, 1, 0);
+		}
 		break;
 
 	case 'k':
-		app->pany += fz_pixmap_height(app->ctx, app->image) / 10;
-		pdfapp_showpage(app, 0, 0, 1, 0);
+		{
+			int image_h = fz_pixmap_height(app->ctx, app->image);
+			if (app->pany == 0 || app->winh >= image_h) {
+				panto = PAN_TO_BOTTOM;
+				app->pageno--;
+				break;
+			}
+			app->pany += image_h / 10;
+			pdfapp_showpage(app, 0, 0, 1, 0);
+		}
 		break;
 
 	case 'l':
@@ -1512,7 +1528,17 @@ void pdfapp_onmouse(pdfapp_t *app, int x, int y, int btn, int modifiers, int sta
 				int isx = (modifiers & (1<<0));
 				int xstep = isx ? 20 * dir : 0;
 				int ystep = !isx ? 20 * dir : 0;
-				pdfapp_panview(app, app->panx + xstep, app->pany + ystep);
+				int image_h = fz_pixmap_height(app->ctx, app->image);
+				if (ystep > 0 && app->pageno > 1 && (app->pany == 0 || app->winh >= image_h)) {
+					app->pany = -2000;
+					app->pageno--;
+					pdfapp_showpage(app, 1, 1, 1, 0);
+				} else if (ystep < 0 && app->pageno < app->pagecount &&
+						(app->pany + ystep + image_h <= app->winh || app->winh >= image_h)) {
+					app->pany = 0;
+					app->pageno++;
+					pdfapp_showpage(app, 1, 1, 1, 0);
+				} else pdfapp_panview(app, app->panx + xstep, app->pany + ystep);
 			}
 		}
 	}
-- 
1.8.3.4

